### PR TITLE
Refactor instance_distance into CorpusAnalyzer subclass GenderDistanceAnalyzer

### PR DESCRIPTION
**Pull request by ryaanahmed**
This PR refactors the various analytic methods in instance_distance into a single CorpusAnalyzer subclass, `GenderDistanceAnalyzer`, which computes the distances between gendered pronouns for each Document in a Corpus and then provides a number of helper methods to organize that data by Document metadata.

This refactor maintains most of the original functionality of the module, but cuts one function `get_highest_distances` which found the documents with the greatest median distance between certain gendered pronouns. This felt like very project-specific code for the Gender Novels project (see the results and use at the bottom of http://gendernovels.digitalhumanitiesmit.org/info/instance_distance_analysis).

We also change the shape of the outputs here: previously the filtering functions returned a shape like this:
```
{
    METADATA_VALUE: {    
        <Document Object>: {
            <Gender Object>: metric_result
        }
    }
}
```
where `metric_result` was one of mean, median, min, or max.

Now they return
```
{
    METADATA_VALUE: {
        <Gender object>: {
            'median': float,
            'mean: float,
            'min': int,
            'max': int
        }
    }
}
```
We remove the intermediary Document object, and these are now _corpus-wide_ by metadata, rather than per-document. We might want to reintroduce the filters so the innermost thing isn't always all of the stats, but rather a specific result. Not sure.

### CODE REVIEW

File: gender_analysis/analysis/common.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+def compute_bin_year(year, time_frame_start, time_frame_end, bin_size):
+    """
+    Given an input year, the start and end of a time frame, and bin size,
+    computes which bin start date this year belongs to.
+    """
+    return ((year - time_frame_start) // bin_size) * bin_size + time_frame_start
```
Comment (MBJean) [inline gender_analysis/analysis/common.py:7]: What's the purpose of the `time_frame_end` argument here? Maybe looks vestigial.

Comment (ryaanahmed) [reply gender_analysis/analysis/common.py:7]: Vestigial: didn't realize when I was writing the function signature that I wouldn't need the ending arg. Will remove.

**Comment by ryaanahmed** [comment]
Coverage failure looks like a Github Actions API failure - not sure what's going on there.

**Review by MBJean — APPROVED** [review]
This LGTM at quick glance. I wonder at the distinction between the `results_by` and `stats_by` methods. Are `results_by` meant to be internal only? Or would it be worth condensing them into single methods with optional parameters? In any case, we can address that later as an optimization (if there's an optimization to make).

**Comment by ryaanahmed** [comment]
> This LGTM at quick glance. I wonder at the distinction between the results_by and stats_by methods. Are results_by meant to be internal only? Or would it be worth condensing them into single methods with optional parameters? In any case, we can address that later as an optimization (if there's an optimization to make).

I wasn't sure: the original versions of all of these returned a single stat, with a filtering argument (so, e.g., you'd request just the means or medians). I thought the aggregated raw results might be useful.

Maybe a better interface is to condense the results_by and stats_by methods into single methods with add an optional argument, specifying whether the method should return the raw results or just one of the statistics. The whole bundle of statistics, as it returns now, is pretty unwieldy. And this would better follow the pattern you setup in #169.

**Pull request merged**